### PR TITLE
feat(presets): pinGitHubActionDigestsToSemver

### PR DESCRIPTION
**Pull request by HonkingGoose**
@coderabbitai: ignore

## Changes

- Create new `helpers:pinGitHubActionDigestsToSemver` preset

## Context

Based on:

- #21901
- #23577

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/helpers.ts:33]: Pinging RegEx wizard @TWiStErRob for a review. 😄

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [inline lib/config/presets/internal/helpers.ts:40]: Do you want to support v1.2.3, v1.2, v1? If yes:
```suggestion
          'regex:^v(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?$',
```
If not:
```suggestion
          'regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)$',
```

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:40]: and the same should probably be done to `extractVersion`?

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:33]: What are the requirements? This looks fairly simple and functional.

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:33]: Ok, that's a very cool feature!

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:40]: Huh, based on https://semver.org/ all 3 components are required, but IRL not all GitHub actions may follow the 3-component versioning.

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:33]: I don't know renovate enough though, so I'll let the maintainers review the functional part. The only comments I have about the regex are below.

**Comment by suzuki-shunsuke** [comment]
Thank you for addressing my feature request!

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:40]: > but IRL not all GitHub actions may follow the 3-component versioning.

Many actions that I've seen have two "version options", `v3` or `v3.0.0`.

- `v3`: give me the latest action in the `v3` range
- `v3.0.0`: give me this specific action, and let me update to a new version manually or via Renovate

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:40]: Yeah, that's normal and *recommended*, but enforced in any way.

Examples:
https://github.com/szenius/set-timezone/tags 
https://github.com/npalm/action-docs-action/tags

These repos have open issues to use the vX tags, but not yet. And the first one uses 2 components only.

Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:40]: Looks like the old "garbage in, garbage out" problem, where we have to work-around projects using silly versions.

I'll let the maintainers decide what to do here.

File: lib/config/presets/internal/helpers.ts
Comment (MaxymVlasov) [reply lib/config/presets/internal/helpers.ts:38]: Inside

https://github.com/renovatebot/.github/blob/95451c7263e7d6ef42198dceb188de271f8bedf0/default.json#L48

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:40]: Considering the name "Pin...ToSemver", requiring 3 makes sense, but following the principle of least astonishment, `3.0` should "just work" the same way as `3.0.0`.

**Comment by HonkingGoose** [comment]
@TWiStErRob can you give me _new_ suggestions to resolve the comments? I don't know which suggestions I should apply... 🙈

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:33]: This is resolved too, it was just a question. For everything else we need maintainer involvement.

File: lib/config/presets/internal/helpers.ts
Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts]: @HonkingGoose you might want to leave this open for visibility of maintainer reviews.

**Review by TWiStErRob — APPROVED** [review]

**Comment by travi** [comment]
originally led to this approach by https://github.com/renovatebot/renovate/pull/23577#discussion_r1278071980, i thought i'd try to provide some thoughts from my experience after [adding the work-around to my personal preset](https://github.com/form8ion/renovate-config/commit/f1bd80c6c0435cbc70f7f03b88a9f015bfdc1663).

> I think only some readers will want this information.

this is from the comment that i linked to above. while i think this specific quote was specifically about including the reference to the workaround alongside the docs for the `pinGitHubActionDigests` helper, i'm calling this out because i think i can really get behind the idea of this being closely related to pinning actions digests. i would maybe even go as far as suggesting that this should be default behavior of pinning digests and that the only reason i can think of for not choosing to do this is not knowing about it.

### My goals with digests

my goal with pinning digests is to make sure that the version of the action referenced in git for a certain commit remains fully repeatable, so that i avoid the situation where a tag could be moved to a different commit. my assumption was that this would still reference specific releases, but adding this rule has helped me recognize that my expectations might not have aligned with the real behaviors. i value the visibility that the version comment adds so that i can understand the current state of things, but i did not expect that the comment might also be driving some behavior. 

### Scenarios with pinned digests

i can think of three scenarios that i've experienced since adding the pinning helper [further back](https://github.com/form8ion/renovate-config/commit/34b06af9bd2c50d1534a66f4550d435f90dd98f6) than this additional rule.

1. digest with a major version comment. often the result of switching to pinning for a version that i referenced the docs of an action to determine the version to reference
2. digest with a full semver comment. the result of the new rule when starting from a digest with a major version
3. digest with no comment. usually an accident. i update my yaml files programmatically often and with an AST tool that doesnt handle comments. i try to revert the dropping of the comments before committing the changes, but sometimes miss catching it

### behavior observations

* one of the things that stood out to me the most after adding this rule was that i started getting release notes included in the PRs after getting to state 2. i had grown to assume that those simply werent available for actions updates, but i think it was more because of being in states 1 or 3.
* it seems like at least state 3 would send PRs for digest changes that dont align with released versions. i imagine these are real commits that dont end up triggering releases because they are changes in the repo that dont have consumer impact. if my understanding is correct. those updates are not valuable to me. i noticed this behavior when adding comments back in after noting they'd been lost in some places to get them back to either state 1 or 2. i was surprised that after adding the latest released version as a comment, i would get a PR realigning the digest with the actual version that was released. this surprised me because i thought the latest digest would have already been aligned with the latest released version.
* it became clear to me that renovate was better at getting the version and comment aligned than i was doing manually (part of the reason i saw value in the pin helper and this new rule in the first place), so it made me wish that renovate would add the comment back automatically in cases where i accidentally dropped it. if there was another config option to enable this behavior, i would absolutely enable it

### collecting my thoughts

sorry for the lengthy context that somewhat hijacks this thread. happy to be redirected elsewhere if it would be valuable, but seemed related enough to this effort that i thought it would be valuable to start here.

what i'm building to is this. i think the addition of this helper will be a great addition. however, i think there might be something lost in thinking about implementing this that might deserve stepping back and considering from a higher level. my goal is to pin the digest for reproducibility and avoid potential attacks because of a tag being moved to a different commit. i still want to see the corresponding version, so the comment is valuable. i want it there always, even if i accidentally remove it. since i'm already pinning to an exact version through the digest, the version comment might as well also be as specific as possible. 

to tie in the [comment from above](https://github.com/renovatebot/renovate/pull/23663#discussion_r1281732079), i want to clarify that i want the pinning that happens to correspond to real _releases_, whatever that means. being fully semver compliant is less valuable than aligning to real releases. i dont know the details of the underlying implementation, but i'd love to see the updates be driven by the version comment rather than the digest. when a release happens, i want the version comment to be updated to the most specific available and the digest to align to that released version.

i think this proposed helper is a big step in the right direction and i'm eager to add it to some other projects, but i'm hoping that stepping back to consider the goal might help think through how to iterate forward.

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [inline lib/config/presets/internal/helpers.ts:33]: Marking this PR as draft, and creating this PR review comment, to make sure we discuss @travi's comment first:

- https://github.com/renovatebot/renovate/pull/23663#issuecomment-1693806214

Comment (travi) [reply lib/config/presets/internal/helpers.ts:33]: to maybe clarify, my goal isnt to hold up this helper from moving forward. i think this helper would be a useful step, especially where i'd like to use it with teams that might not grasp the goal with the existing raw workaround. 

instead, i was just hoping to share some thoughts about where follow-up steps could maybe take this idea a bit further if the bigger picture could be captured. i thought that might also help to give some potentially different context to consider the unresolved questions from the other comment thread above.

**Comment by HonkingGoose** [comment]
## Make pin to SemVer default behavior?

> this is from the comment that i linked to above. while i think this specific quote was specifically about including the reference to the workaround alongside the docs for the `pinGitHubActionDigests` helper, i'm calling this out because i think i can really get behind the idea of this being closely related to pinning actions digests. i would maybe even go as far as suggesting that this should be default behavior of pinning digests and that the only reason i can think of for not choosing to do this is not knowing about it.

@travi Yes, that comment was about putting a load of text into a `description` field. The preset docs page is meant to only show short descriptions. And text in a `description` is harder to style properly.

@travi It sounds like you want to change the `helpers:pinGitHubActionDigests` preset, so it pins digests to their SemVer version, by default? So that users don't have to opt-in to another preset to get the actual best-practice behavior?

Here's the current code for the `helpers:pinGitHubActionDigest` preset:

```json
{
  "packageRules": [
    {
      "matchDepTypes": [
        "action"
      ],
      "pinDigests": true
    }
  ]
}
```

### About digest pinning

> ### My goals with digests
> 
> my goal with pinning digests is to make sure that the version of the action referenced in git for a certain commit remains fully repeatable, so that i avoid the situation where a tag could be moved to a different commit. my assumption was that this would still reference specific releases, but adding this rule has helped me recognize that my expectations might not have aligned with the real behaviors. i value the visibility that the version comment adds so that i can understand the current state of things, but i did not expect that the comment might also be driving some behavior.

GitHub recommend that you pin third-party actions to a specific commit, to prevent bad actors from making a malicious release. As an added bonus, pinning to a specific commit also makes your CI more reproducible.

### About the three states in your comment

I'll let one of the maintainer deal with this part, I'm not qualified to talk about the behavior or the code. 😄 

### About your behavior observations

> it became clear to me that renovate was better at getting the version and comment aligned than i was doing manually (part of the reason i saw value in the pin helper and this new rule in the first place), so it made me wish that renovate would add the comment back automatically in cases where i accidentally dropped it. if there was another config option to enable this behavior, i would absolutely enable it

I guess Renovate doesn't have enough information based on just the commit hash what it should do:

1. Add comment to follow full SemVer version
2. Add comment to follow partial version like `@v3`
3. Do nothing
4. Let user configure "missing comment behavior" with a new config option `whatToDoIfNoCommentFoundInGitHubActionFile`
5. Something else

I don't know what happens if Renovate sees a commit hash without the tag comment, right now. I found some relevant text in the `github-action` manager docs [^1]. I've **bolded** the key part:

> The `github-actions` manager extracts dependencies from GitHub Actions workflow and workflow template files.
> 
> If you like to use digest pinning but want to follow the action version tag, you can use the following sample:
> 
> ```yaml
> name: build
> 
> on: [push]
> 
> jobs:
>   build:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@af513c7a016048ae468971c52ed77d9562c7c819 # v1.0.0
> ```
> 
> **Renovate will update the commit SHA but follow the GitHub tag you specified.**
> Renovate can update digests that use SHA1 and SHA256 algorithms.
> 
> If you want to automatically pin action digests add the `helpers:pinGitHubActionDigests` preset to the `extends` array:
> 
> ```json
> {
>   "extends": ["helpers:pinGitHubActionDigests"]
> }
> ```

## About collecting my thoughts

> my goal is to pin the digest for reproducibility and avoid potential attacks because of a tag being moved to a different commit. i still want to see the corresponding version, so the comment is valuable. i want it there always, even if i accidentally remove it. since i'm already pinning to an exact version through the digest, the version comment might as well also be as specific as possible.

Agreed.

> to tie in the https://github.com/renovatebot/renovate/pull/23663#discussion_r1281732079, i want to clarify that i want the pinning that happens to correspond to real releases, whatever that means. being fully semver compliant is less valuable than aligning to real releases. i dont know the details of the underlying implementation, but i'd love to see the updates be driven by the version comment rather than the digest. when a release happens, i want the version comment to be updated to the most specific available and the digest to align to that released version.

I like your point of aliging to "real releases"!

### HonkingGoose's summary of your idea

Checking in to see if I understand your comment correctly. I think this is the short summary of what you want?

1. Renovate to replace any missing comments.
2. Renovate to align to "real releases".
3. Renovate to always pin to the full SemVer version.
4. Renovate to keep you safe from malicious "release bad version on same Git tag" stuff

Please correct me if I'm wrong! 😉 

[^1]: https://docs.renovatebot.com/modules/manager/github-actions/#additional-information

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:33]: Maybe we should hold up this helper... 😉

I think the big point is that we should make Renovate easy to use, and easy to do the right thing. So we should think if adding a new preset is really the right thing to do. 😄 

Maybe making a breaking change to the `helpers:pinGitHubActionDigests` preset in `v37`, so it always "pins GitHub Actions digests to SemVer versions" is better than making a new `pinGitHubActionDigestsToSemverPreset`. I'd prefer for Renovate to just do the right thing out of the box. 😄 That way I don't have to search through a lot of config options to find what I want. 🙃

Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:33]: I also found another issue that seems related:

- https://github.com/renovatebot/renovate/issues/7516

**Comment by rarkins** [comment]
What's an example change which this would make? Can someone give me an example of before and after?

**Comment by travi** [comment]
> It sounds like you want to change the `helpers:pinGitHubActionDigests` preset, so it pins digests to their SemVer version, by default? So that users don't have to opt-in to another preset to get the actual best-practice behavior?

yes, i think there would be value in enabling this behavior by default. it does feel like a best-practice to me. 

the one detail that i would clarify here is that i think the semver precision is more of a side-effect than the target behavior. i know this is repetitive with the similar point in the "collecting my thoughts" section, so this is just reinforcing that point. i dont fully understand the relationships that renovate is capable of managing, but it seems to me that aligning the comment to the most specific _release_ version available is the best target regardless of whether that is fully semver or not. it is common for github actions to have to tracked versions, one with just the major (vx) and one that is a full semver (vx.y.z). however, some projects choose other patterns like major (vx) and vx.y. in a case like that, using the vx.y is valuable, even though it isnt a full vx.y.z. 

> Checking in to see if I understand your comment correctly. I think this is the short summary of what you want?
> 
> 1. Renovate to replace any missing comments.
> 2. Renovate to align to "real releases".
> 3. Renovate to always pin to the full SemVer version.
> 4. Renovate to keep you safe from malicious "release bad version on same Git tag" stuff
> 
> Please correct me if I'm wrong! 😉

#3 is the only one that i'd again possibly question the wording of. like above, i think i'd land somewhere closer to "most specific version available from a real release". i think we are basically saying the same thing, but i think there are real situations where a full semver version doesnt exist in a real release, so i think this rewording mostly helps to define the behavior in the less common scenarios where a full semver release doesnt exist


overall, i think you've captured my point well

**Comment by travi** [comment]
> What's an example change which this would make? Can someone give me an example of before and after?

if i'm understanding your question correctly, these are examples of the proposed preset making the version comment more specific: 
* https://github.com/form8ion/project/commit/5e37ee050b77cb9be2620d0ae84b161bbd104e10
* https://github.com/form8ion/project/commit/a36a50c93362c189da7cd7d0ef14978894c90eb1

in case it is helpful, this is an example of manually re-adding an accidentally removed version comment and the steps renovate currently follows to get back to a good state:
* [the re-add of the version comment(s)](https://github.com/form8ion/meta/commit/c9bfb4450d43427da209f7aa9cc2388c9e6f172b)
* [get the digest back in sync with the major version tag](https://github.com/form8ion/meta/commit/df0d6c94325c3c93e01e84504882753c735f03cb) (i'm fairly confident this was because the digest matched a commit that was ahead of the position of the v3 tag)
* [update the version comment to match the more specific version](https://github.com/form8ion/meta/commit/79e3532a73f279aa2be6bd5516a39f4cd1c9c9f5)

do these examples clarify what you were asking about?

**Comment by rarkins** [comment]
Thanks, that helps to clarify. The 3 -> 3.1.4 part is what we'd normally refer to as pinning _versions_, so the "digests" part in the title here threw me off (and maybe needs reconsideration).

Let's consider two similar scenarios:
1. You have no SHA pinned digest, and use a "range" version like v3
2. Same as (a) but you have a SHA (and the version is in a comment like `# v3`)

For (1), going from v3 to v3.1.4 does make a difference to future behavior. With v3 alone, you wouldn't get any PR until a v4 is one day released. For (2) you'll get PRs any time the v3 tag is updated to point to a different SHA, which is quite likely to be any time a new semver release comes out (e.g. v3.1.5). So in (2), you are already going to get a Renovate PR approximately every time there's a new semver release anyway and pinning to v3.1.4 does not cause more noise. On the upside, it will allow you to get release notes and more transparency about what version you're using.

In other words:
- If no SHA exists, and pinDigests=false, leave it alone
- If a SHA exists, but no version comment, then add it
- If a SHA exists and version comment, but it's not pinned to semver, then pin it

Is the above aligned with your thinking too?

**Comment by travi** [comment]
i appreciate the terminology clarifications. those terms align with my understanding, but i was failing to articulate the same way :)

> In other words:
> 
> * If no SHA exists, and pinDigests=false, leave it alone
> * If a SHA exists, but no version comment, then add it
> * If a SHA exists and version comment, but it's not pinned to semver, then pin it
> 
> Is the above aligned with your thinking too?

yes, this would align well with my goals and feels like it captures what i would consider to be the best practice for the situation.

> you are already going to get a Renovate PR approximately every time there's a new semver release anyway and pinning to v3.1.4 does not cause more noise

agreed

> On the upside, it will allow you to get release notes and more transparency about what version you're using.

release notes are always valuable, but also since the sha is already fully specific, the transparency is valuable and feels like it would be the intent of most folks wanting to pin

> For (1), going from v3 to v3.1.4 does make a difference to future behavior. With v3 alone, you wouldn't get any PR until a v4 is one day released.

i appreciate that you called out this example. i think this being available highlights that there would still be an option available to use a version range, if desired. with that available, it feels even more appropriate to me that when pinning, all the things should be pinned without a need to make pinning more configurable.

**Comment by HonkingGoose** [comment]
So what do we want to do now? It looks like we've discussed most of the requirements and nice to haves above? We should probably make new issues to split up the different parts of the discussion. 😄 

Heads-up, we already have one issue that's related to the discussion above:

- https://github.com/renovatebot/renovate/issues/7516

File: lib/config/presets/internal/helpers.ts
Comment (rarkins) [reply lib/config/presets/internal/helpers.ts]: Let's go with 3 components

**Comment by rarkins** [comment]
My preferred functionality (copy pasted from earlier) is this:

- If no SHA exists, and pinDigests=false, leave it alone
- If a SHA exists, but no version comment, then add it
- If a SHA exists and version comment, but it's not pinned to semver, then pin it

I'm surprised and happy if this can already be done using a _preset_ alone

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts]: @TWiStErRob do you want to suggest the correct regexes for 3 component versions? 🙃

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts]: 😂, just remove the question-mark and then the redundant parentheses, and unit test!

Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts]: Oh I already applied your suggestion:

- https://github.com/renovatebot/renovate/pull/23663/commits/938a4f5515751705ef3664369a85ea81bd837fa3

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts]: Yeah reverse that one, because Rhys wants only 1 and 3 components only.

**Comment by rarkins** [comment]
What's the next steps here?

**Comment by HonkingGoose** [comment]
> What's the next steps here?

Somebody should check if the regex that's currently in my PR will do this:

> My preferred functionality (copy pasted from earlier) is this:
> 
>     * If no SHA exists, and pinDigests=false, leave it alone
> 
>     * If a SHA exists, but no version comment, then add it
> 
>     * If a SHA exists and version comment, but it's not pinned to semver, then pin it
> 
> 
> I'm surprised and happy if this can already be done using a _preset_ alone

**Review by rarkins — COMMENTED** [review]
Can some examples of this in action be shared?

**Comment by HonkingGoose** [comment]
> Can some examples of this in action be shared?

@rarkins I see that `@travi` tried some code on their own repository:

- https://github.com/semantic-release/.github/pull/20

**Comment by TWiStErRob** [comment]
@rarkins, I think you're looking for these: [org:semantic-release is:pr created:2023-10-23 "update actions"](https://github.com/search?q=org%3Asemantic-release+is%3Apr+created%3A2023-10-23+%22update+actions%22&type=pullrequests&s=created&o=desc).
The PR mentioned by @HonkingGoose was merged 23rd 18:46 PM. Major updates (e.g. node v4) seem to contain the version number before the change too, but for example these two look like the consequence of this change:
 * https://github.com/semantic-release/github/pull/726
 * https://github.com/semantic-release/github/pull/725
Weird thing is that this is also after the merge:
 * https://github.com/semantic-release/github/pull/724

**Review by rarkins — APPROVED** [review]
I'm not sure exactly how/why this works, but it's opt-in/optional for users so I'm happy to merge it and let people try it

File: lib/config/presets/internal/helpers.ts
Comment (kristof-mattei) [reply lib/config/presets/internal/helpers.ts:38]: This rule breaks actions like `rui314/setup-mold@v1`. They can no longer be pinned as there is no `v1.0.0` available at all.

Proposal to change this to:

`      "extractVersion": "^(?<version>v\\d+(?:\\.\\d+){0,2})$",`

That way `actions/checkout@v4` resolves all the way to `v4.1.1` and becomes pinned to that. (no change)

And for `rui314/setup-mold@v1` we detect `v1` as a valid version and allow pinning on that.

Example Repo: 

https://github.com/kristof-mattei/renovate-mold-no-version

PR to pin digests: https://github.com/kristof-mattei/renovate-mold-no-version/pull/4

PR that still updates actions/checkout@v4 to v4.1.1:

https://github.com/kristof-mattei/renovate-mold-no-version/pull/5

Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:38]: Copy/pasting the proposal from @kristof-mattei so the maintainers and @TWiStErRob can review it:

```suggestion
        extractVersion: '^(?<version>v\\d+(?:\\.\\d+){0,2})$',
```

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:38]: LGTM syntactically and it seems meaningful.

I might ask consider simplifying it with capturing groups (no `?:`) and simple optional sequences `(...)?` just to use less regex features, which should equate to more people understanding it easier. Iff this doesn't break things for extraction.

How does this relate to https://github.com/renovatebot/renovate/pull/23663#discussion_r1281723709 where an executive decision was made for not supporting 2 components?

Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:38]: > I might ask consider simplifying it with capturing groups (no `?:`) and simple optional sequences `(...)?` just to use less regex features, which should equate to more people understanding it easier. Iff this doesn't break things for extraction.

Can you suggest the correct regex for your idea?

I'll let @rarkins comment on the "executive decision they made about only supporting 3 version actions".

Comment (TWiStErRob) [reply lib/config/presets/internal/helpers.ts:38]: For 1-2-3 components:
```suggestion
        extractVersion: '^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$',
```
For 1-3 components:
```suggestion
        extractVersion: '^(?<version>v\\d+(\\.\\d+\\.\\d+)?)$',
```

Comment (kristof-mattei) [reply lib/config/presets/internal/helpers.ts:38]: @HonkingGoose I'll add one thing as to why I believe it makes sense to adopt this PR, but maybe under a different name.

I agree that v1 is not a semver version.

However, the reason I proposed this is because I had issues with `rui314/setup-mold` which only has a `v1` tag.

It seems that applying this ruleset disables the ability of the github-actions updater to consider versions like `v1`.

I believe that, even though `v1` isn't semver compatible, it is still valuable for me to be bale to pin `rui314@setup-mold@v1` to `rui314@setup-mold@sha-1234 #v1` AND `actions/checkout@v4` to `actions/checkout@sha-5678 # v4.1.1`

For the latter nothing changes.

But for the former now I can track the changes with both the sha & tag.

**Review by rarkins — APPROVED** [review]

File: lib/config/presets/internal/helpers.ts
Comment (HonkingGoose) [reply lib/config/presets/internal/helpers.ts:38]: @rarkins can you review this review thread and decide what you want to do? 😉

Comment (rarkins) [reply lib/config/presets/internal/helpers.ts:38]: It doesn't seem right to refer to it as "semver" if it's supporting short tags too. Is this then simply a "pin to longest version" PR, or is it more complicated than that?

Also keep in mind that this will be an opt-in preset so it doesn't _break_ anything, it simply may not solve all scenarios

Comment (kristof-mattei) [reply lib/config/presets/internal/helpers.ts:38]: > It doesn't seem right to refer to it as "semver" if it's supporting short tags too. Is this then simply a "pin to longest version" PR, or is it more complicated than that?
> 
> Also keep in mind that this will be an opt-in preset so it doesn't _break_ anything, it simply may not solve all scenarios

You're right. My proposal clashes with the name. 

I'll propose a separate `pinToMostDetailedVersion`.

**Comment by MPV** [comment]
What remains to resolve this? 
Looks like "only" an unresolved conversation to me.
@TWiStErRob would you be okay with this PR now (given that you raised the still open discussions)?

File: lib/config/presets/internal/helpers.ts
Comment (MPV) [reply lib/config/presets/internal/helpers.ts:38]: FYI I haven't gotten the above to work — i.e. I tried this in my `renovate.json`:
```diff
-  }
+  },
+  "packageRules": [
+    {
+      "description": "When pinning GitHub Action digests, use most detailed version",
+      "matchDepTypes": ["action"],
+      "extractVersion": "^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$",
+      "versioning": "regex:^v(?<major>\\d+)(\\.(?<minor>\\d+)\\.(?<patch>\\d+))?$"
+    }
+  ]
 }
```
...and no improvement after letting Renovate retry/rebase my PRs.

Maybe I must use the `extends: ['helpers:pinGitHubActionDigests'],` for this to work at all — would they conflict?

Comment (MPV) [reply lib/config/presets/internal/helpers.ts:38]: Ah, I just realized I had picked the "wrong" version, changing it now (in my repo's renovate.json) with:
```diff
-      "extractVersion": "^(?<version>v\\d+(\\.\\d+)?(\\.\\d+)?)$",
+      "extractVersion": "^(?<version>v\\d+\\.\\d+\\.\\d+)$",
```

**Comment by TWiStErRob** [comment]
@MPV that mega-thread is probably 3 conversations away from the original which was "resolved" in August: https://github.com/renovatebot/renovate/pull/23663/files#r1299844474, only collaborators and PR author can resolve conversations. I'm lost in all the follow-ups :)

Ignoring all the convos, the code make sense in the PR. I disagree with with only supporting vx and vx.y.z, but that was the maintainer's decision, so I have to accept.

**Review by rarkins — APPROVED** [review]

**Pull request merged**
